src/OSSupport/Network.cpp

```diff
@@ -0,0 +1,1259 @@
+	virtual UInt16 GetLocalPort(void) const override { return m_LocalPort; }
+	virtual AString GetRemoteIP(void) const override { return m_RemoteIP; }
+	virtual UInt16 GetRemotePort(void) const override { return m_RemotePort; }
+	virtual void Shutdown(void) override;
+	virtual void Close(void) override;
+
+protected:
+
+	/** Callbacks to call when the connection is established.
+	May be NULL if not used. Only used for outgoing connections (cNetwork::Connect()). */
+	cNetwork::cConnectCallbacksPtr m_ConnectCallbacks;
+
+	/** The LibEvent handle representing this connection. */
+	bufferevent * m_BufferEvent;
+
+	/** The server handle that has created this link.
+	Only valid for incoming connections, NULL for outgoing connections. */
+	cServerHandleImpl * m_Server;
+
+	/** The IP address of the local endpoint. Valid only after the socket has been connected. */
+	AString m_LocalIP;
+
+	/** The port of the local endpoint. Valid only after the socket has been connected. */
+	UInt16 m_LocalPort;
+
+	/** The IP address of the remote endpoint. Valid only after the socket has been connected. */
+	AString m_RemoteIP;
+
+	/** The port of the remote endpoint. Valid only after the socket has been connected. */
+	UInt16 m_RemotePort;
+
+
+	/** Callback that LibEvent calls when there's data available from the remote peer. */
+	static void ReadCallback(bufferevent * a_BufferEvent, void * a_Self);
+
+	/** Callback that LibEvent calls when there's a non-data-related event on the socket. */
+	static void EventCallback(bufferevent * a_BufferEvent, short a_What, void * a_Self);
+
+	/** Sets a_IP and a_Port to values read from a_Address, based on the correct address family. */
+	static void UpdateAddress(const sockaddr * a_Address, int a_AddrLen, AString & a_IP, UInt16 & a_Port);
+
+	/** Updates m_LocalIP and m_LocalPort based on the metadata read from the socket. */
+	void UpdateLocalAddress(void);
+
+	/** Updates m_RemoteIP and m_RemotePort based on the metadata read from the socket. */
+	void UpdateRemoteAddress(void);
+};
+typedef SharedPtr<cTCPLinkImpl> cTCPLinkImplPtr;
+typedef std::vector<cTCPLinkImplPtr> cTCPLinkImplPtrs;
+
+
+
+
+
+/** Implements the cServerHandle details so that it can represent a real server socket, with a list of clients. */
+class cServerHandleImpl:
+	public cServerHandle
+{
+	typedef cServerHandle super;
+	friend class cTCPLinkImpl;
+
+public:
+	/** Creates a new instance with the specified callbacks.
+	Initializes the internals, but doesn't start listening yet. */
+	cServerHandleImpl(
+		cNetwork::cListenCallbacksPtr a_ListenCallbacks,
+		cTCPLink::cCallbacksPtr a_LinkCallbacks
+	);
+
+	/** Closes the server, dropping all the connections. */
+	~cServerHandleImpl();
+
+	/** Starts listening on the specified port.
+	Both IPv4 and IPv6 interfaces are used, if possible. */
+	bool Listen(UInt16 a_Port);
+
+	// cServerHandle overrides:
+	virtual void Close(void) override;
+	virtual bool IsListening(void) const override { return m_IsListening; }
```

Comment:
When the `listen` syscall fails (most likely another socket already listening on that port), we still need to return a valid `cServerHandle` instance (so that the Lua API can read the error code and message from it).